composer/utils/batch_helpers.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Callable, Optional, Sequence, Union
+
+__all__ = ['batch_get', 'batch_set']
+
+
+def batch_get(batch: Any, key: Optional[Any] = None, get_fn: Optional[Callable[[Any], Any]] = None) -> Any:
+    """Indexes into the batch given the key.
+
+    >>> from composer.utils.batch_helpers import batch_get
+    >>> batch_get([1,2,3], 1)
+    2
+    >>> batch_get({'a':1, 'b':7}, 'b')
+    7
+    >>> batch_get([{'a':1, 'b':7},{'c':5}], get_fn=lambda x: x[1]['c'])
+    5
+
+    Args:
+        batch (Any): An object that contains the input and label of the items in the batch.
+            Can be any abritrary type that user creates, but we assume some sort of
+            sequence (list, tuple, tensor, array), mapping (dictionary),
+            or attribute store (object with data members, namedtuple).
+        key (Any): A key to index into the batch. Key is optional if get_fn is supplied.
+        get_fn (Callable): A user-specified function to do the extracting. 
+        get_fn is optional if key is supplied.
+
+    Returns:
+        The part of the batch specified by the key or the get_fn. This could be any type 
+            depending on what the batch is composed of.
+    
+    Raises:
+        ValueError if key is unset and get_fn is unset or if both are set.
+    """
+    if key is None and get_fn is None:
+        raise ValueError("key or get_fn must be specified and neither were!")
+    if key is not None and get_fn is not None:
+        raise ValueError("key and get_fn were both set. Only one can be set!")
+    if get_fn is not None:
+        return get_fn(batch)
+    if isinstance(key, Sequence) and not isinstance(key, str):
+        return _batch_get_multiple(batch, key)
+    else:
+        return _batch_get(batch, key)
+
+
+def _batch_get(batch: Any, key: Any) -> Any:
+    if isinstance(batch, tuple):
+        return _batch_get_tuple(batch, key)
+
+    exceptions = []
+    try:
+        return batch[key]
+
+    # The only acceptable TypeError is for an object that doesn't have a __getitem__,
+    # which is TypeError("... object is not subscriptable").
+    except TypeError as e:
+        if 'object is not subscriptable' in str(e):
+            exceptions.append(e)
+            pass
+        else:
+            raise e
+
+    try:
+        return getattr(batch, key)
+
+    # If both getattr and __getitem__ result in exceptions then raise both of them.
+    except (AttributeError, TypeError) as e:
+        exceptions.append(e)
+        raise Exception(exceptions)
```

Comment:
See comment below on the set_fn exception.

Reply:
done here: 
bd3dce1